pkg/fqdn/dnsproxy/proxy_test.go

```diff
@@ -524,62 +540,286 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+
+	// Case 2 | EPID1 | dstIP1 |   54 | cilium.io      | Rejected | Port 54 only allows example.com
+	allowed, err = s.proxy.CheckAllowed(epID1, 54, 2, dstIP1, "cilium.io")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
+
+	// Case 3 | EPID1 | dstIP2a |   53 | cilium.io      | Allowed
+	allowed, err = s.proxy.CheckAllowed(epID1, 53, 2, dstIP2a, "cilium.io")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, true, Commentf("request was rejected when it should be allowed"))
+
+	// Case 4 | EPID1 | dstIP2b |   53 | aws.amazon.com | Rejected | Only cilium.io is allowed with DstID2
+	allowed, err = s.proxy.CheckAllowed(epID1, 53, 2, dstIP2b, "aws.amazon.com")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
+
+	// Case 5 | EPID1 | dstIP1 |   54 | example.com    | Allowed
+	allowed, err = s.proxy.CheckAllowed(epID1, 54, 2, dstIP1, "example.com")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, true, Commentf("request was rejected when it should be allowed"))
+
+	// make sure random IP is not allowed
+	// Case 5 | EPID1 | random IP |   54 | example.com    | Rejected
+	allowed, err = s.proxy.CheckAllowed(epID1, 54, 2, dstIPrandom, "example.com")
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
+
+	// Restore rules for epID3
+	s.proxy.RestoreRules(uint16(epID3), restored3)
+	_, exists = s.proxy.restored[epID3]
+	c.Assert(exists, Equals, true)
+
+	// Set empty ruleset, check that restored rules were deleted in epID3
+	err = s.proxy.UpdateAllowed(epID3, 53, nil)
+	c.Assert(err, Equals, nil, Commentf("Could not update with rules"))
+
+	_, exists = s.proxy.restored[epID3]
+	c.Assert(exists, Equals, false)
+
+	// epID1 still has restored rules
+	_, exists = s.proxy.restored[epID1]
+	c.Assert(exists, Equals, true)
+
+	// Marshal restored rules to JSON
+	jsn, err := json.Marshal(s.proxy.restored[epID1])
+	c.Assert(err, Equals, nil, Commentf("Could not marshal restored rules to json"))
+
+	expected := `
+	{
+		"53": {
+			"(^[-a-zA-Z0-9_]*[.]ubuntu[.]com[.]$)|(^aws[.]amazon[.]com[.]$)": {"::": {}},
+			"(^cilium[.]io[.]$)": {"127.0.0.1": {}, "127.0.0.2": {}}
+		},
+		"54": {
+			"(^example[.]com[.]$)": {"::": {}}
+		}
+	}`
+	pretty := new(bytes.Buffer)
+	err = json.Compact(pretty, []byte(expected))
+	c.Assert(err, Equals, nil, Commentf("Could not compact expected json"))
+	c.Assert(string(jsn), Equals, pretty.String())
+
+	s.proxy.RemoveRestoredRules(uint16(epID1))
+	_, exists = s.proxy.restored[epID1]
+	c.Assert(exists, Equals, false)
+
+	// Before restore: everything is dropped
```

Comment:
What's the difference between this section and the section from lines 646-708?

Reply:
before the earlier block all normal rules were removed. Here we just removed the current restored rules. It that removal did not work, we might not notice json marshal/unmarshal problems in the following tests.

Reply:
I changed the comments of these two test blocks to add context.